### DISCUSSION TITLE
batch normalization in Unet

### DISCUSSION THREAD

**Discussion by anais2390** [Project-MONAI/MONAI#3909] [Q&A]
Hi everyone !

I would like to add batch normalization in monai `UNet` but I can't figure out how. I suppose I should use the argument `norm` but I don't understand how ? I tried `norm=Norm.BATCH` but this didn't work.

Can anyone give me an exemple ? 
Thanks you for your help !

**Comment by wyli** [comment]
could you give more details, I had a quick test it seems the `norm` works fine with the latest codebase:

```py
>>> from monai.networks.layers.factories import Act, Norm
>>> from monai.networks.nets import UNet
>>> net = UNet(2, 1, 1, (4,8,16), (2, 2), norm=Norm.BATCH)
>>> net(torch.zeros(1, 1, 64, 64))
tensor([[[[-0.3094, -0.3094, -0.3094,  ..., -0.3094, -0.3094, -0.3094],
          [-0.3094, -0.3094, -0.3094,  ..., -0.3094, -0.3094, -0.3094],
          [-0.3094, -0.3094, -0.3094,  ..., -0.3094, -0.3094, -0.3094],
          ...,
          [-0.3094, -0.3094, -0.3094,  ..., -0.3094, -0.3094, -0.3094],
          [-0.3094, -0.3094, -0.3094,  ..., -0.3094, -0.3094, -0.3094],
          [-0.3094, -0.3094, -0.3094,  ..., -0.3094, -0.3094, -0.3094]]]],
       grad_fn=<SlowConvTranspose2DBackward0>)
>>> 
```

**Reply by anais2390** [reply]
Hi @wyli, thanks a lot for your answer.

Actually I don't have any error when running UNet training with `norm=Norm.BATCH` but my training curves don't show any convergence, whereas using the norm default value, I have a nice convergence of the loss and MeanDice (I am running a segmentation training).

I run the test several times with and without `norm=Norm.BATCH` and I had every times, the same results.

It seemed very unlikely that the batch normalization would degrade the results that much, so I thought I had set it up wrong.

What do you think ?

**Reply by wyli** [reply]
I think maybe the batchnorm doesn't work, you can try other normalization with additional parameters such as 
```py
net = UNet(2, 1, 1, (4, 8, 16), (2, 2), norm=(Norm.GROUP, {'num_groups': 2, 'affine': False}))
```

**Reply by anais2390** [reply]
I'll try that, thanks @wyli .

This is confusing me, I though batch normalization was a sure option that was worth using for any problematic.

Could you explain how the Norm.GROUP is working ?

**Reply by wyli** [reply]
no problem, internally it's a thin wrapper of [`torch.nn.GroupNorm`](https://pytorch.org/docs/stable/generated/torch.nn.GroupNorm.html) via:
https://github.com/Project-MONAI/MONAI/blob/4b66746764e200a9a933d2198a7aee13e53b1206/monai/networks/layers/utils.py#L28-L31
and
https://github.com/Project-MONAI/MONAI/blob/4b66746764e200a9a933d2198a7aee13e53b1206/monai/networks/layers/factories.py#L227
